File: leetcode-04/solutions/cpp/0718.cpp
class Solution {
 public:
  int findLength(vector<int>& A, vector<int>& B) {
    const int m = A.size();
    const int n = B.size();

    int ans = 0;
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));

    for (int i = m - 1; i >= 0; i--)
      for (int j = n - 1; j >= 0; j--)
        if (A[i] == B[j]) dp[i][j] = dp[i + 1][j + 1] + 1;

    for (vector<int>& row : dp)
      ans = max(ans, *max_element(row.begin(), row.end()));

    return ans;
  }
};

File: leetcode-04/solutions/python3/0718.py
class Solution:
    def findLength(self, A: List[int], B: List[int]) -> int:
        m = len(A)
        n = len(B)

        dp = [[0] * (n + 1) for _ in range(m + 1)]

        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if A[i] == B[j]:
                    dp[i][j] = dp[i + 1][j + 1] + 1

        return max(max(row) for row in dp)
